Applications/Spire/Source/Ui/ScrollArea.cpp

```diff
@@ -0,0 +1,165 @@
+      setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+    } else if(is_within_vertical_scroll_bar(e->pos().x()) &&
+        !horizontalScrollBar()->isSliderDown()) {
+      set_scroll_bar_style(SCROLL_BAR_MAX_SIZE);
+      setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOn);
+      setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+    } else {
+      set_scroll_bar_style(SCROLL_BAR_MIN_SIZE);
+      if(!m_v_scroll_bar_timer.isActive() &&
+          verticalScrollBarPolicy() != Qt::ScrollBarAlwaysOff &&
+          !verticalScrollBar()->isSliderDown()) {
+        hide_vertical_scroll_bar();
+      }
+      if(!m_h_scroll_bar_timer.isActive() &&
+          horizontalScrollBarPolicy() != Qt::ScrollBarAlwaysOff &&
+          !horizontalScrollBar()->isSliderDown()) {
+        hide_horizontal_scroll_bar();
+      }
+    }
+  } else if(event->type() == QEvent::HoverLeave) {
+    if(!m_h_scroll_bar_timer.isActive() &&
+        horizontalScrollBarPolicy() == Qt::ScrollBarAlwaysOff) {
+      hide_horizontal_scroll_bar();
+    }
+    if(!m_v_scroll_bar_timer.isActive() &&
+        verticalScrollBarPolicy() == Qt::ScrollBarAlwaysOff) {
+      hide_vertical_scroll_bar();
+    }
+  }
+  return QScrollArea::eventFilter(watched, event);
+}
+
+void ScrollArea::leaveEvent(QEvent* event) {
+  hide_horizontal_scroll_bar();
+  hide_vertical_scroll_bar();
+}
+
+void ScrollArea::wheelEvent(QWheelEvent* event) {
+  if(event->modifiers() & Qt::ShiftModifier) {
+    setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+    setHorizontalScrollBarPolicy(Qt::ScrollBarAsNeeded);
+    horizontalScrollBar()->setValue(horizontalScrollBar()->value() -
+      (event->delta() / 2));
+    m_h_scroll_bar_timer.start();
+  } else if(!event->modifiers().testFlag(Qt::ShiftModifier)) {
+    setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+    setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);
+    verticalScrollBar()->setValue(verticalScrollBar()->value() -
+      (event->delta() / 2));
+    m_v_scroll_bar_timer.start();
+  }
+}
+
+void ScrollArea::hide_horizontal_scroll_bar() {
+  m_h_scroll_bar_timer.stop();
+  setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+}
+
+void ScrollArea::hide_vertical_scroll_bar() {
+  m_v_scroll_bar_timer.stop();
+  setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+}
+
+bool ScrollArea::is_within_horizontal_scroll_bar(int pos_y) {
+  if(widget() == nullptr) {
+    return false;
+  }
+  auto bar = verticalScrollBar();
+  auto scroll_adj = map_to(static_cast<double>(bar->value()),
+    static_cast<double>(bar->minimum()), static_cast<double>(bar->maximum()),
+    0, widget()->height() - height());
+  return pos_y - scroll_adj > height() - scale_height(SCROLL_BAR_MAX_SIZE);
+}
+
+bool ScrollArea::is_within_vertical_scroll_bar(int pos_x) {
+  if(widget() == nullptr) {
+    return false;
+  }
+  auto bar = horizontalScrollBar();
+  auto scroll_adj = map_to(static_cast<double>(bar->value()),
```

Comment:
Consider factoring this mapping routine out as the base is identical to `is_within_horizontal_scroll_bar`.